Sources/System/FilePath/FilePathSyntax.swift

```diff
@@ -0,0 +1,353 @@
+  ///
+  /// If `basename` is `nil` or one of the special path components
+  /// `.` or `..`, `get` returns `nil` and `set` does nothing.
+  ///
+  /// If `basename` does not contain a `.` anywhere, or only
+  /// at the start, `get` returns `nil` and `set` will append a
+  /// `.` and `newValue` to `basename`
+  ///
+  /// Otherwise `get` returns everyting after it, and `set` will
+  /// replace the extension
+  ///
+  /// Examples:
+  ///   * `/tmp/foo.txt => "txt"`
+  ///   * `/Appliations/Foo.app/ => app`
+  ///   * `/Appliations/Foo.app/bar.txt => txt`
+  ///   * `/tmp/.hidden => nil``
+  ///   * `/tmp/.. => nil``
+  ///
+  /// REVIEW NOTE: Rust has a set_extension instead, which returns whether it
+  /// did anything, but having a setter is more convenient.
+  public var `extension`: String? {
+    get {
+      fatalError()
+    }
+    set { fatalError() }
+  }
+
+  /// The non-extension portion of the file or directoy last component.
+  ///
+  /// Returns `nil` if `basename` is `nil`
+  ///
+  ///   * `/tmp/foo.txt => "foo"`
+  ///   * `/Appliations/Foo.app/ => Foo`
+  ///   * `/Appliations/Foo.app/bar.txt => bar`
+  ///   * `/tmp/.hidden => .hidden``
+  ///   * `/tmp/.. => ..``
+  ///   * `/ => nil`
+  ///
+  /// TODO: setter? That's a little harder to imagine than for `extension`.
+  /// Setting to `nil` is more likely a bug.
+  public var stem: String? {
+    fatalError()
+  }
+
+}
+
+// Modification and concatenation API
+extension FilePath {
+  /// Remove the contents, keeping the null termiantor
+  ///
+  /// TODO: better name? I like `clear` but that's not "Swifty"
+  /// TODO: Since we'll still have the null terminator, should we intern the empty path?
+  public mutating func removeContents(keepingCapacity: Bool = false) {
+    fatalError()
+  }
+
+  /// Reserve enough storage space to store `minimumCapacity` path characters
+  public mutating func reserveCapacity(_ minimumCapacity: Int) {
+    self.storage.reserveCapacity(1 + minimumCapacity)
+  }
+
+  /// Append the contents of `other`.
+  ///
+  /// TODO: What if `other` is absolute? C++ does replacement for append
+  /// but not for concat... Rust's `push` does replacement. Should we just do push/pop?
+  ///
+  public mutating func append(_ other: FilePath) {
+    // TODO: Faster to do byte copy, checking for leading separator
+    self.components.append(contentsOf: other.components)
+  }
+
+  /// Remove and return the last component of this file path. If the path is
+  /// root or empty, does nothing and returns `nil`.
+  ///
+  /// Examples:
+  /// * `"/".pop() == nil          // path is /`
+  /// * `"/foo/bar".pop() == "bar" // path is /foo`
+  public mutating func pop() -> FilePath.Component? {
+    self.components.popLast() // FIXME: popLastButNotRoot()
+  }
+
+  /// See `append`
+  ///
+  /// TODO: Should this be `/`? Should the RHS have to be Component?
+  public static func +(_ lhs: FilePath, _ rhs: FilePath) -> FilePath {
```

Comment:
I think `+` is the right choice here.  One concern with the operator combined with the string literal conversion is that it isn't obvious whether each `foo + "usr" + "local" + "include"` does, based on whether `foo` is a path or string; in that case it looks a lot like string concatenation.  But `/` doesn't feel right since it's specific to the Unix form of the path.

If the RHS has to be a component, then there would need to be some other way to append a path.

In TSC's `AbsolutePath` and `RelativePath` we avoided the `+` operator for this reason.  Instead there is `AbsolutePath.append(_ subpath: RelativePath)` and `AbsolutePath.append(component: String)` (as well as `AbsolutePath.append(components: String...)`), but that does get a bit cumbersome to type.

Reply:
I think if we do an operator it will be `/` instead of `+`, since we won't be modeling "concatenation" in a stringy sense. C++ uses `/` and it seems to be clear and natural even on Windows (where `/` can be used as a path separator prior to normalization). Not using an operator is also fine.